src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,231 @@
+#include "table_scan.hpp"
+#include "get_table.hpp"
+#include "resolve_type.hpp"
+#include "storage/abstract_attribute_vector.hpp"
+#include "storage/table.hpp"
+#include "storage/value_segment.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+TableScan::TableScan(const std::shared_ptr<const AbstractOperator>& in, const ColumnID column_id,
+                     const ScanType scan_type, const AllTypeVariant search_value)
+    : AbstractOperator{in}, _column_id{column_id}, _scan_type{scan_type}, _search_value{search_value} {}
+
+ColumnID TableScan::column_id() const {
+  return _column_id;
+}
+
+ScanType TableScan::scan_type() const {
+  return _scan_type;
+}
+
+const AllTypeVariant& TableScan::search_value() const {
+  return _search_value;
+}
+
+std::shared_ptr<const Table> TableScan::_on_execute() {
+  const auto input_table = _left_input_table();
+
+  Assert(input_table, "Performing a table scan without input does not work.");
+  const auto chunk_count = input_table->chunk_count();
+  const auto column_type = input_table->column_type(_column_id);
+  auto output_reference_segments = std::vector<std::shared_ptr<ReferenceSegment>>{};
+
+  // any comparison with null will always return an empty set
+  if (variant_is_null(_search_value)) {
+    return std::make_shared<Table>(*input_table, output_reference_segments);
+  }
+
+  // We might end up using far less segments, but still should be worth to reserve the space for worst case.
+  output_reference_segments.reserve(chunk_count);
+
+  resolve_data_type(column_type, [&](auto type) {
+    using Type = typename decltype(type)::type;
+
+    for (auto chunk_id = ChunkID{0}; chunk_id < chunk_count; ++chunk_id) {
+      const auto chunk = input_table->get_chunk(chunk_id);
+      // if chunk is empty, skip this chunk
+      if (!chunk->size()) {
+        continue;
+      }
+
+      auto position_list = std::make_shared<PosList>();
+      const auto segment = chunk->get_segment(_column_id);
+
+      const auto value_segment = std::dynamic_pointer_cast<ValueSegment<Type>>(segment);
+      const auto dictionary_segment = std::dynamic_pointer_cast<DictionarySegment<Type>>(segment);
+      const auto reference_segment = std::dynamic_pointer_cast<ReferenceSegment>(segment);
+
+      Assert(value_segment || dictionary_segment || reference_segment,
+             "TableScan was called on unsupported segment type");
+
+      if (value_segment) {
+        position_list = _tablescan_value_segment<Type>(value_segment, chunk_id);
+      } else if (dictionary_segment) {
+        position_list = _tablescan_dict_segment<Type>(dictionary_segment, chunk_id);
+      } else if (reference_segment) {
+        position_list = _tablescan_reference_segment<Type>(reference_segment, chunk_id);
+      }
```

Comment:
You could do the cast in the if-statements like following:
```suggestion

      if (auto value_segment = std::dynamic_pointer_cast<ValueSegment<Type>>(segment); value_segment) {
        position_list = _tablescan_value_segment<Type>(value_segment, chunk_id);
      } else if ( auto dictionary_segment = std::dynamic_pointer_cast<DictionarySegment<Type>>(segment); dictionary_segment) {
        position_list = _tablescan_dict_segment<Type>(dictionary_segment, chunk_id);
      } else if (auto reference_segment = std::dynamic_pointer_cast<ReferenceSegment>(segment); reference_segment) {
        position_list = _tablescan_reference_segment<Type>(reference_segment, chunk_id);
      }
      else {
            Fail( "TableScan was called on unsupported segment type");
      }
```